File: go/format.go
package counters

import (
	"math/big"
	"strconv"
	"strings"
)

// FormatOptions carries the optional settings for FormatCompact, FormatScientific, and
// DescribeMagnitude. A nil *FormatOptions, or a nil MaxFractionDigits field, selects the
// default of 3 fraction digits. FormatFull accepts no options.
type FormatOptions struct {
	// MaxFractionDigits bounds the fractional digits in a rendered quotient: an integer from
	// 0 through 20. Any other value is a validation error, reported even when the value would
	// not display a fraction.
	MaxFractionDigits *int
}

const (
	defaultMaxFractionDigits = 3
	minMaxFractionDigits     = 0
	maxMaxFractionDigits     = 20
)

// formatLadderStep is one power-of-ten divisor and the label appended at that scale.
type formatLadderStep struct {
	divisor *big.Int
	label   string
}

// roundedFormatQuotient is a quotient rendered as scaledValue / scale: exact integer output
// of the shared HALF_UP procedure, with scale always a power of ten.
type roundedFormatQuotient struct {
	scale       *big.Int
	scaledValue *big.Int
}

// normalizedFormatValue is a validated input split into sign and absolute value, with the
// absolute value's canonical decimal digits cached for digit counting and grouping.
type normalizedFormatValue struct {
	absolute *big.Int
	digits   string
	negative bool
}

var (
	formatThousand = big.NewInt(1000)
	formatTen      = big.NewInt(10)
	formatTwo      = big.NewInt(2)

	compactLadder = []formatLadderStep{
		{formatPowerOfTen(3), "K"},
		{formatPowerOfTen(6), "M"},
		{formatPowerOfTen(9), "B"},
		{formatPowerOfTen(12), "T"},
		{formatPowerOfTen(15), "P"},
		{formatPowerOfTen(18), "E"},
		{formatPowerOfTen(21), "Z"},
		{formatPowerOfTen(24), "Y"},
		{formatPowerOfTen(27), "R"},
		{formatPowerOfTen(30), "Q"},
	}

	magnitudeLadder = []formatLadderStep{
		{formatPowerOfTen(3), "Thousand"},
		{formatPowerOfTen(6), "Million"},
		{formatPowerOfTen(9), "Billion"},
		{formatPowerOfTen(12), "Trillion"},
		{formatPowerOfTen(15), "Quadrillion"},
		{formatPowerOfTen(18), "Quintillion"},
		{formatPowerOfTen(21), "Sextillion"},
		{formatPowerOfTen(24), "Septillion"},
		{formatPowerOfTen(27), "Octillion"},
		{formatPowerOfTen(30), "Nonillion"},
		{formatPowerOfTen(33), "Decillion"},
	}

	compactScientificThreshold   = formatPowerOfTen(33)
	magnitudeScientificThreshold = formatPowerOfTen(36)
)

// FormatCompact renders value with an SI-style compact suffix (K, M, B, …). Values at or
// beyond the top of the suffix ladder fall back to scientific notation. All arithmetic is
// exact math/big integer work; the output is byte-identical to the other SDKs.
func FormatCompact(value string, opts *FormatOptions) (string, error) {
	maxFractionDigits, err := validateFormatOptions(opts)
	if err != nil {
		return "", err
	}
	normalized, err := normalizeFormatString(value)
	if err != nil {
		return "", err
	}
	formatted := formatLadderAbsolute(normalized.absolute, maxFractionDigits, compactLadder, "", compactScientificThreshold)
	return applyFormatSign(formatted, normalized.negative), nil
}

// FormatCompactBig is FormatCompact for a *big.Int value.
func FormatCompactBig(value *big.Int, opts *FormatOptions) (string, error) {
	maxFractionDigits, err := validateFormatOptions(opts)
	if err != nil {
		return "", err
	}
	normalized, err := normalizeFormatBig(value)
	if err != nil {
		return "", err
	}
	formatted := formatLadderAbsolute(normalized.absolute, maxFractionDigits, compactLadder, "", compactScientificThreshold)
	return applyFormatSign(formatted, normalized.negative), nil
}

// FormatScientific renders value with one leading mantissa digit, a lowercase e, and a
// base-ten exponent. Values below 10 render as the plain integer; zero renders as "0".
func FormatScientific(value string, opts *FormatOptions) (string, error) {
	maxFractionDigits, err := validateFormatOptions(opts)
	if err != nil {
		return "", err
	}
	normalized, err := normalizeFormatString(value)
	if err != nil {
		return "", err
	}
	formatted := formatScientificAbsolute(normalized.absolute, maxFractionDigits)
	return applyFormatSign(formatted, normalized.negative), nil
}

// FormatScientificBig is FormatScientific for a *big.Int value.
func FormatScientificBig(value *big.Int, opts *FormatOptions) (string, error) {
	maxFractionDigits, err := validateFormatOptions(opts)
	if err != nil {
		return "", err
	}
	normalized, err := normalizeFormatBig(value)
	if err != nil {
		return "", err
	}
	formatted := formatScientificAbsolute(normalized.absolute, maxFractionDigits)
	return applyFormatSign(formatted, normalized.negative), nil
}

// FormatFull renders every exact digit of value, grouped by thousands with ASCII commas.
// There is no rounding and no length limit.
func FormatFull(value string) (string, error) {
	normalized, err := normalizeFormatString(value)
	if err != nil {
		return "", err
	}
	return applyFormatSign(groupFormatDigits(normalized.digits), normalized.negative), nil
}

// FormatFullBig is FormatFull for a *big.Int value.
func FormatFullBig(value *big.Int) (string, error) {
	normalized, err := normalizeFormatBig(value)
	if err != nil {
		return "", err
	}
	return applyFormatSign(groupFormatDigits(normalized.digits), normalized.negative), nil
}

// DescribeMagnitude renders value with a short-scale English magnitude word (Thousand,
// Million, …). Values at or beyond the top of the word ladder fall back to scientific
// notation.
func DescribeMagnitude(value string, opts *FormatOptions) (string, error) {
	maxFractionDigits, err := validateFormatOptions(opts)
	if err != nil {
		return "", err
	}
	normalized, err := normalizeFormatString(value)
	if err != nil {
		return "", err
	}
	formatted := formatLadderAbsolute(normalized.absolute, maxFractionDigits, magnitudeLadder, " ", magnitudeScientificThreshold)
	return applyFormatSign(formatted, normalized.negative), nil
}

// DescribeMagnitudeBig is DescribeMagnitude for a *big.Int value.
func DescribeMagnitudeBig(value *big.Int, opts *FormatOptions) (string, error) {
	maxFractionDigits, err := validateFormatOptions(opts)
	if err != nil {
		return "", err
	}
	normalized, err := normalizeFormatBig(value)
	if err != nil {
		return "", err
	}
	formatted := formatLadderAbsolute(normalized.absolute, maxFractionDigits, magnitudeLadder, " ", magnitudeScientificThreshold)
	return applyFormatSign(formatted, normalized.negative), nil
}

func validateFormatOptions(opts *FormatOptions) (int, error) {
	if opts == nil || opts.MaxFractionDigits == nil {
		return defaultMaxFractionDigits, nil
	}
	maxFractionDigits := *opts.MaxFractionDigits
	if maxFractionDigits < minMaxFractionDigits || maxFractionDigits > maxMaxFractionDigits {
		return 0, &ValidationError{"maxFractionDigits must be an integer from 0 through 20"}
	}
	return maxFractionDigits, nil
}

func normalizeFormatString(value string) (*normalizedFormatValue, error) {
	if !valueRe.MatchString(value) {
		return nil, &ValidationError{"value must be a signed integer: " + value}
	}
	n, ok := new(big.Int).SetString(value, 10)
	if !ok {
		return nil, &ValidationError{"value must be a signed integer: " + value}
	}
	return splitFormatSign(n), nil
}

func normalizeFormatBig(value *big.Int) (*normalizedFormatValue, error) {
	if value == nil {
		return nil, &ValidationError{"value must be non-nil"}
	}
	return splitFormatSign(value), nil
}

// splitFormatSign separates sign from magnitude so rounding happens on the absolute value
// and HALF_UP ties round away from zero for negative inputs.
func splitFormatSign(n *big.Int) *normalizedFormatValue {
	negative := n.Sign() < 0
	absolute := new(big.Int).Abs(n)
	return &normalizedFormatValue{
		absolute: absolute,
		digits:   absolute.Text(10),
		negative: negative,
	}
}

func applyFormatSign(formattedAbsolute string, negative bool) string {
	if negative {
		return "-" + formattedAbsolute
	}
	return formattedAbsolute
}

func formatLadderAbsolute(absolute *big.Int, maxFractionDigits int, ladder []formatLadderStep, separator string, scientificThreshold *big.Int) string {
	if absolute.Cmp(formatThousand) < 0 {
		return absolute.Text(10)
	}
	if absolute.Cmp(scientificThreshold) >= 0 {
		return formatScientificAbsolute(absolute, maxFractionDigits)
	}

	stepIndex := findFormatLadderStep(absolute, ladder)
	step := ladder[stepIndex]
	rounded := roundFormatQuotient(absolute, step.divisor, maxFractionDigits)

	// A quotient that rounds to 1000 belongs to the next ladder step.
	if rounded.scaledValue.Cmp(new(big.Int).Mul(formatThousand, rounded.scale)) >= 0 {
		stepIndex++
		if stepIndex >= len(ladder) {
			return formatScientificAbsolute(absolute, maxFractionDigits)
		}
		step = ladder[stepIndex]
		rounded = roundFormatQuotient(absolute, step.divisor, maxFractionDigits)
	}

	return renderFormatScaledInteger(rounded) + separator + step.label
}

func findFormatLadderStep(absolute *big.Int, ladder []formatLadderStep) int {
	for index := len(ladder) - 1; index >= 0; index-- {
		if absolute.Cmp(ladder[index].divisor) >= 0 {
			return index
		}
	}

	// Callers handle values below the first step before searching the ladder.
	panic("ladder has no matching step")
}

func formatScientificAbsolute(absolute *big.Int, maxFractionDigits int) string {
	if absolute.Cmp(formatTen) < 0 {
		return absolute.Text(10)
	}

	// The exponent is a decimal digit count: bounded metadata, never the value itself.
	exponent := len(absolute.Text(10)) - 1
	divisor := formatPowerOfTen(exponent)
	rounded := roundFormatQuotient(absolute, divisor, maxFractionDigits)

	// Rounding 9.99… to 10 requires a one-place exponent promotion.
	if rounded.scaledValue.Cmp(new(big.Int).Mul(formatTen, rounded.scale)) >= 0 {
		exponent++
		rounded = roundedFormatQuotient{
			scale:       rounded.scale,
			scaledValue: new(big.Int).Set(rounded.scale),
		}
	}

	return renderFormatScaledInteger(rounded) + "e" + strconv.Itoa(exponent)
}

func roundFormatQuotient(absolute, divisor *big.Int, maxFractionDigits int) roundedFormatQuotient {
	scale := formatPowerOfTen(maxFractionDigits)
	scaledDividend := new(big.Int).Mul(absolute, scale)
	scaledValue, remainder := new(big.Int), new(big.Int)
	// Every operand is non-negative, so truncated division is floor division here.
	scaledValue.QuoRem(scaledDividend, divisor, remainder)

	// Exact HALF_UP: an exact half increments the positive scaled quotient.
	if new(big.Int).Mul(formatTwo, remainder).Cmp(divisor) >= 0 {
		scaledValue.Add(scaledValue, big.NewInt(1))
	}

	return roundedFormatQuotient{scale: scale, scaledValue: scaledValue}
}

func renderFormatScaledInteger(rounded roundedFormatQuotient) string {
	fractionDigits := len(rounded.scale.Text(10)) - 1
	if fractionDigits == 0 {
		return rounded.scaledValue.Text(10)
	}

	digits := rounded.scaledValue.Text(10)
	if len(digits) < fractionDigits+1 {
		digits = strings.Repeat("0", fractionDigits+1-len(digits)) + digits
	}
	wholeEnd := len(digits) - fractionDigits
	whole := digits[:wholeEnd]
	fraction := strings.TrimRight(digits[wholeEnd:], "0")
	if fraction == "" {
		return whole
	}
	return whole + "." + fraction
}

func formatPowerOfTen(exponent int) *big.Int {
	return new(big.Int).Exp(formatTen, big.NewInt(int64(exponent)), nil)
}

func groupFormatDigits(digits string) string {
	if len(digits) <= 3 {
		return digits
	}

	firstGroupLength := len(digits) % 3
	if firstGroupLength == 0 {
		firstGroupLength = 3
	}
	groups := []string{digits[:firstGroupLength]}
	for index := firstGroupLength; index < len(digits); index += 3 {
		groups = append(groups, digits[index:index+3])
	}
	return strings.Join(groups, ",")
}

File: typescript/src/format.ts
import { CountersValidationError } from "./errors.js";

export interface FormatOptions {
  maxFractionDigits?: number;
}

interface NormalizedValue {
  absolute: bigint;
  digits: string;
  negative: boolean;
}

interface LadderStep {
  divisor: bigint;
  label: string;
}

interface RoundedQuotient {
  scale: bigint;
  scaledValue: bigint;
}

const SIGNED_INTEGER_RE = /^-?[0-9]+$/;
const TRAILING_ZERO_RE = /0+$/;
const DEFAULT_MAX_FRACTION_DIGITS = 3;
const MIN_MAX_FRACTION_DIGITS = 0;
const MAX_MAX_FRACTION_DIGITS = 20;
const THOUSAND = 1000n;

const COMPACT_LADDER: readonly LadderStep[] = [
  { divisor: 10n ** 3n, label: "K" },
  { divisor: 10n ** 6n, label: "M" },
  { divisor: 10n ** 9n, label: "B" },
  { divisor: 10n ** 12n, label: "T" },
  { divisor: 10n ** 15n, label: "P" },
  { divisor: 10n ** 18n, label: "E" },
  { divisor: 10n ** 21n, label: "Z" },
  { divisor: 10n ** 24n, label: "Y" },
  { divisor: 10n ** 27n, label: "R" },
  { divisor: 10n ** 30n, label: "Q" },
];

const MAGNITUDE_LADDER: readonly LadderStep[] = [
  { divisor: 10n ** 3n, label: "Thousand" },
  { divisor: 10n ** 6n, label: "Million" },
  { divisor: 10n ** 9n, label: "Billion" },
  { divisor: 10n ** 12n, label: "Trillion" },
  { divisor: 10n ** 15n, label: "Quadrillion" },
  { divisor: 10n ** 18n, label: "Quintillion" },
  { divisor: 10n ** 21n, label: "Sextillion" },
  { divisor: 10n ** 24n, label: "Septillion" },
  { divisor: 10n ** 27n, label: "Octillion" },
  { divisor: 10n ** 30n, label: "Nonillion" },
  { divisor: 10n ** 33n, label: "Decillion" },
];

const COMPACT_SCIENTIFIC_THRESHOLD = 10n ** 33n;
const MAGNITUDE_SCIENTIFIC_THRESHOLD = 10n ** 36n;

/** Render an integer with an SI-style compact suffix. */
export function formatCompact(value: string | bigint, options?: FormatOptions): string {
  const maxFractionDigits = validateOptions(options);
  const normalized = normalizeValue(value);
  const formatted = formatLadderAbsolute(
    normalized.absolute,
    maxFractionDigits,
    COMPACT_LADDER,
    "",
    COMPACT_SCIENTIFIC_THRESHOLD,
  );
  return applySign(formatted, normalized.negative);
}

/** Render an integer with one leading mantissa digit and a base-ten exponent. */
export function formatScientific(value: string | bigint, options?: FormatOptions): string {
  const maxFractionDigits = validateOptions(options);
  const normalized = normalizeValue(value);
  const formatted = formatScientificAbsolute(normalized.absolute, maxFractionDigits);
  return applySign(formatted, normalized.negative);
}

/** Render every exact digit, grouped by thousands with ASCII commas. */
export function formatFull(value: string | bigint): string {
  const normalized = normalizeValue(value);
  const formatted = groupDigits(normalized.digits);
  return applySign(formatted, normalized.negative);
}

/** Render an integer with a short-scale English magnitude word. */
export function describeMagnitude(value: string | bigint, options?: FormatOptions): string {
  const maxFractionDigits = validateOptions(options);
  const normalized = normalizeValue(value);
  const formatted = formatLadderAbsolute(
    normalized.absolute,
    maxFractionDigits,
    MAGNITUDE_LADDER,
    " ",
    MAGNITUDE_SCIENTIFIC_THRESHOLD,
  );
  return applySign(formatted, normalized.negative);
}

function validateOptions(options: FormatOptions | undefined): number {
  if (options === undefined) {
    return DEFAULT_MAX_FRACTION_DIGITS;
  }
  if (typeof options !== "object" || options === null || Array.isArray(options)) {
    throw new CountersValidationError("format options must be an object");
  }
  if (options.maxFractionDigits === undefined) {
    return DEFAULT_MAX_FRACTION_DIGITS;
  }

  const maxFractionDigits = options.maxFractionDigits;
  if (
    !Number.isInteger(maxFractionDigits) ||
    maxFractionDigits < MIN_MAX_FRACTION_DIGITS ||
    maxFractionDigits > MAX_MAX_FRACTION_DIGITS
  ) {
    throw new CountersValidationError("maxFractionDigits must be an integer from 0 through 20");
  }
  return maxFractionDigits;
}

function normalizeValue(value: string | bigint): NormalizedValue {
  if (typeof value !== "string" && typeof value !== "bigint") {
    throw new CountersValidationError("value must be a bigint or signed integer string");
  }
  if (typeof value === "string" && !SIGNED_INTEGER_RE.test(value)) {
    throw new CountersValidationError("value string must match ^-?[0-9]+$");
  }

  const integer = typeof value === "bigint" ? value : BigInt(value);
  const negative = integer < 0n;
  const absolute = negative ? -integer : integer;
  return {
    absolute,
    digits: absolute.toString(),
    negative,
  };
}

function applySign(formattedAbsolute: string, negative: boolean): string {
  return negative ? `-${formattedAbsolute}` : formattedAbsolute;
}

function formatLadderAbsolute(
  absolute: bigint,
  maxFractionDigits: number,
  ladder: readonly LadderStep[],
  separator: string,
  scientificThreshold: bigint,
): string {
  if (absolute < THOUSAND) {
    return absolute.toString();
  }
  if (absolute >= scientificThreshold) {
    return formatScientificAbsolute(absolute, maxFractionDigits);
  }

  let stepIndex = findLadderStep(absolute, ladder);
  let step = ladder[stepIndex]!;
  let rounded = roundQuotient(absolute, step.divisor, maxFractionDigits);

  // A quotient that rounds to 1000 belongs to the next ladder step.
  if (rounded.scaledValue >= THOUSAND * rounded.scale) {
    stepIndex += 1;
    if (stepIndex >= ladder.length) {
      return formatScientificAbsolute(absolute, maxFractionDigits);
    }
    step = ladder[stepIndex]!;
    rounded = roundQuotient(absolute, step.divisor, maxFractionDigits);
  }

  return `${renderScaledInteger(rounded)}${separator}${step.label}`;
}

function findLadderStep(absolute: bigint, ladder: readonly LadderStep[]): number {
  for (let index = ladder.length - 1; index >= 0; index -= 1) {
    if (absolute >= ladder[index]!.divisor) {
      return index;
    }
  }

  // Callers handle values below the first step before searching the ladder.
  throw new Error("ladder has no matching step");
}

function formatScientificAbsolute(absolute: bigint, maxFractionDigits: number): string {
  if (absolute < 10n) {
    return absolute.toString();
  }

  let exponent = absolute.toString().length - 1;
  const divisor = powerOfTen(exponent);
  let rounded = roundQuotient(absolute, divisor, maxFractionDigits);

  // Rounding 9.99… to 10 requires a one-place exponent promotion.
  if (rounded.scaledValue >= 10n * rounded.scale) {
    exponent += 1;
    rounded = {
      scale: rounded.scale,
      scaledValue: rounded.scale,
    };
  }

  return `${renderScaledInteger(rounded)}e${exponent}`;
}

function roundQuotient(
  absolute: bigint,
  divisor: bigint,
  maxFractionDigits: number,
): RoundedQuotient {
  const scale = powerOfTen(maxFractionDigits);
  const scaledDividend = absolute * scale;
  let scaledValue = scaledDividend / divisor;
  const remainder = scaledDividend % divisor;

  // Exact HALF_UP: an exact half increments the positive scaled quotient.
  if (2n * remainder >= divisor) {
    scaledValue += 1n;
  }

  return { scale, scaledValue };
}

function renderScaledInteger(rounded: RoundedQuotient): string {
  const fractionDigits = rounded.scale.toString().length - 1;
  if (fractionDigits === 0) {
    return rounded.scaledValue.toString();
  }

  const digits = rounded.scaledValue.toString().padStart(fractionDigits + 1, "0");
  const wholeEnd = digits.length - fractionDigits;
  const whole = digits.slice(0, wholeEnd);
  const fraction = digits.slice(wholeEnd).replace(TRAILING_ZERO_RE, "");
  return fraction.length === 0 ? whole : `${whole}.${fraction}`;
}

function powerOfTen(exponent: number): bigint {
  return 10n ** BigInt(exponent);
}

function groupDigits(digits: string): string {
  if (digits.length <= 3) {
    return digits;
  }

  const firstGroupLength = digits.length % 3 || 3;
  const groups = [digits.slice(0, firstGroupLength)];
  for (let index = firstGroupLength; index < digits.length; index += 3) {
    groups.push(digits.slice(index, index + 3));
  }
  return groups.join(",");
}
